packages/webpack-plugin/src/plugin-utils.ts

```diff
@@ -0,0 +1,368 @@
+    assetsModules,
+    chunkGraph,
+    moduleGraph,
+    runtime,
+    runtimeTemplate,
+    dependencyTemplates,
+}: ReplaceMappedCSSAssetPlaceholdersOptions) {
+    return replaceCSSAssetPlaceholders(
+        stylableBuildMeta,
+        staticPublicPath,
+        (resourcePath, publicPath) => {
+            const assetModule = assetsModules.get(resourcePath);
+            if (!assetModule) {
+                throw new Error('Missing asset module for ' + resourcePath);
+            }
+            if (isLoadedWithKnownAssetLoader(assetModule)) {
+                return extractFilenameFromAssetModule(assetModule, publicPath);
+            } else {
+                const assetModuleSource = assetModule.generator.generate(assetModule, {
+                    chunkGraph,
+                    moduleGraph,
+                    runtime,
+                    runtimeRequirements: new Set(),
+                    runtimeTemplate,
+                    dependencyTemplates,
+                    type: 'asset/resource',
+                });
+
+                if (assetModule.buildInfo.dataUrl) {
+                    return extractDataUrlFromAssetModuleSource(
+                        assetModuleSource.source().toString()
+                    );
+                }
+
+                return publicPath + assetModule.buildInfo.filename;
+            }
+        }
+    );
+}
+
+export function extractFilenameFromAssetModule(m: NormalModule, publicPath: string): string {
+    const source = m.originalSource()!.source();
+    let match = source.toString().match(/__webpack_public_path__\s*\+\s*"(.*?)"/);
+    if (match) {
+        return publicPath + match[1];
+    }
+    match = source.toString().match(/module.exports\s*=\s*"(.*?)"/);
+    if (match) {
+        return match[1];
+    }
+    match = source.toString().match(/export\s+default\s+"(.*?)"/);
+    if (match) {
+        return match[1];
+    }
+    throw new Error('unknown asset module format ' + source);
+}
+
+export function extractDataUrlFromAssetModuleSource(source: string): string {
+    let match = source.match(/.exports\s*=\s*"(.*?)"/);
+    if (match) {
+        return match[1];
+    }
+    match = source.toString().match(/export\s+default\s+"(.*?)"/);
+    if (match) {
+        return match[1];
+    }
+    throw new Error('unknown data url asset module format ' + source);
+}
+
+type AssetNormalModule = NormalModule & { loaders: [{ loader: 'file-loader' | 'url-loader' }] };
+
+export function isLoadedWithKnownAssetLoader(module: Module): module is AssetNormalModule {
+    if (module instanceof NormalModule) {
+        return !!module.loaders.find(({ loader }) =>
+            /[\\/](file-loader)|(url-loader)[\\/]/.test(loader)
+        );
+    }
+    return false;
+}
+
+export function outputOptionsAwareHashContent(
+    createHash: webpackCreateHash,
+    outputOptions: webpackOutputOptions,
+    content: string
+) {
+    const hash = createHash(outputOptions.hashFunction || 'md4');
```

Comment:
md4 as a fallback in 2021? can we use something less prone to conflicts?

Reply:
it is the default from webpack